pandas/tests/frame/test_alter_axes.py

```diff
@@ -28,244 +25,287 @@
+    def test_set_index_pass_single_array(self, drop, append, index_name, box):
+        df = self.dummy.copy()
+        df.index.name = index_name
+
+        # update constructor in case of MultiIndex
+        box = ((lambda x: MultiIndex.from_arrays([x]))
+               if box == 'MultiIndex' else box)
+
+        key = box(df['B'])
+        # np.array and list "forget" the name of B
+        name = [None if box in [np.array, list] else 'B']
+
+        result = df.set_index(key, drop=drop, append=append)
+
+        # only valid column keys are dropped
+        # since B is always passed as array above, nothing is dropped
+        expected = df.set_index(['B'], drop=False, append=append)
+        expected.index.names = [index_name] + name if append else name
+
+        tm.assert_frame_equal(result, expected)
+
+    # also test index name if append=True (name is duplicate here for A & B)
+    @pytest.mark.parametrize('box', [Series, Index, np.array,
+                                     list, 'MultiIndex'])
+    @pytest.mark.parametrize('append, index_name',
+                             [(True, None), (True, 'A'), (True, 'B'),
+                              (True, 'test'), (False, None)])
+    @pytest.mark.parametrize('drop', [True, False])
+    def test_set_index_pass_arrays(self, drop, append, index_name, box):
+        df = self.dummy.copy()
+        df.index.name = index_name
+
+        # update constructor in case of MultiIndex
+        box = ((lambda x: MultiIndex.from_arrays([x]))
+               if box == 'MultiIndex' else box)
+
+        keys = ['A', box(df['B'])]
+        # np.array and list "forget" the name of B
+        names = ['A', None if box in [np.array, list] else 'B']
+
+        result = df.set_index(keys, drop=drop, append=append)
+
+        # only valid column keys are dropped
+        # since B is always passed as array above, only A is dropped, if at all
+        expected = df.set_index(['A', 'B'], drop=False, append=append)
+        expected = expected.drop('A', axis=1) if drop else expected
+        expected.index.names = [index_name] + names if append else names
+
+        tm.assert_frame_equal(result, expected)
+
+    # also test index name if append=True (name is duplicate here for A)
+    @pytest.mark.parametrize('box1', ['label', Series, Index, np.array,
+                                      list, 'MultiIndex'])
+    @pytest.mark.parametrize('box2', ['label', Series, Index, np.array,
+                                      list, 'MultiIndex'])
+    @pytest.mark.parametrize('append, index_name', [(True, None),
+                             (True, 'A'), (True, 'test'), (False, None)])
+    @pytest.mark.parametrize('drop', [True, False])
+    def test_set_index_pass_arrays_duplicate(self, drop, append, index_name,
+                                             box1, box2):
+        df = self.dummy.copy()
+        df.index.name = index_name
+
+        # transform strings to correct box constructor
+        def rebox(x):
```

Comment:
this is not good at all, pls don't do this inside the test function, just use a lambda in the box itself